File: dotnet/Allmark/RenderConsole/ConsoleLinkRenderer.cs
namespace Allmark.Render;

using Allmark.Types;

public static class ConsoleLinkRenderer
{
    public static Renderer Create()
    {
        return new Renderer
        {
            Name = "link",
            Render = (node, state, first, last, decode) => Render(node, state),
        };
    }

    public static void Render(MarkdownNode node, RendererState state)
    {
        var style = Ansi.Blue + Ansi.Underline;
        state.Output.Append(style);
        RenderChildren.Execute(node, state);
        if (!string.IsNullOrEmpty(node.Info))
        {
            state.Output.Append($"{Ansi.Reset} {Ansi.Dim}({node.Info}){Ansi.Reset}");
        }
        else
        {
            state.Output.Append(Ansi.Reset);
        }
    }
}

File: zig/src/render-console/consoleLinkRenderer.zig
const std = @import("std");

const MarkdownNode = @import("../types/MarkdownNode.zig").MarkdownNode;
const ConsoleRendererState = @import("../types/RendererState.zig").RendererState;
const Renderer = @import("../types/Renderer.zig").Renderer;
const ansiBlue = @import("./renderToConsole.zig").ansiBlue;
const ansiUnderline = @import("./renderToConsole.zig").ansiUnderline;
const ansiDim = @import("./renderToConsole.zig").ansiDim;
const ansiReset = @import("./renderToConsole.zig").ansiReset;
const renderChildrenConsole = @import("./renderToConsole.zig").renderChildrenConsole;

pub const consoleLinkRenderer = Renderer{
    .name = "link",
    .render = render,
};

pub fn render(node: *const MarkdownNode, state: *ConsoleRendererState, first: ?bool, last: ?bool, decode: ?bool) void {
    _ = first;
    _ = last;
    _ = decode;

    state.output.appendSlice(state.allocator, ansiUnderline ++ ansiBlue) catch unreachable;
    renderChildrenConsole(node, state, true) catch unreachable;
    state.output.appendSlice(state.allocator, ansiReset) catch unreachable;

    if (node.info) |info| {
        state.output.append(state.allocator, ' ') catch unreachable;
        state.output.appendSlice(state.allocator, ansiDim) catch unreachable;
        state.output.append(state.allocator, '(') catch unreachable;
        state.output.appendSlice(state.allocator, info) catch unreachable;
        state.output.append(state.allocator, ')') catch unreachable;
        state.output.appendSlice(state.allocator, ansiReset) catch unreachable;
    }
}
